debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/outbox/MongoEventRouter.java

```diff
@@ -0,0 +1,337 @@
+import io.debezium.data.Envelope;
+import io.debezium.time.Timestamp;
+import io.debezium.transforms.outbox.EventRouterConfigDefinition;
+import io.debezium.transforms.outbox.EventRouterDelegate;
+
+/**
+ * Debezium MongoDB Outbox Event Router SMT
+ *
+ * @author Sungho Hwang
+ */
+public class MongoEventRouter<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MongoEventRouter.class);
+
+    private final JsonWriterSettings jsonWriterSettings = JsonWriterSettings.builder()
+            .outputMode(JsonMode.EXTENDED)
+            .indent(true)
+            .newLineCharacters("\n")
+            .build();
+
+    private final MongoDataConverter converter = new MongoDataConverter(ExtractNewDocumentState.ArrayEncoding.ARRAY);
+
+    private String fieldTimestamp;
+    private String fieldPayload;
+    private boolean expandPayload;
+
+    private final ExtractField<R> afterExtractor = new ExtractField.Value<>();
+    private final EventRouterDelegate<R> eventRouterDelegate = new EventRouterDelegate<>();
+
+    @Override
+    public R apply(R r) {
+        return eventRouterDelegate.apply(r, (record) -> {
+            try {
+                return expandAfterField(r);
+            }
+            catch (Exception e) {
+                LOGGER.warn("Failed to expand after field: " + e.getMessage(), e);
+                return r;
+            }
+        });
+    }
+
+    @Override
+    public ConfigDef config() {
+        return MongoEventRouterConfigDefinition.configDef();
+    }
+
+    @Override
+    public void close() {
+        eventRouterDelegate.close();
+    }
+
+    @Override
+    public void configure(Map<String, ?> configMap) {
+        final Configuration config = Configuration.from(configMap);
+        fieldTimestamp = config.getString(MongoEventRouterConfigDefinition.FIELD_EVENT_TIMESTAMP);
+        expandPayload = config.getBoolean(MongoEventRouterConfigDefinition.EXPAND_JSON_PAYLOAD);
+        fieldPayload = config.getString(MongoEventRouterConfigDefinition.FIELD_PAYLOAD);
+
+        final Map<String, String> afterExtractorConfig = new HashMap<>();
+        afterExtractorConfig.put("field", Envelope.FieldName.AFTER);
+
+        afterExtractor.configure(afterExtractorConfig);
+
+        // Convert configuration fields from MongoDB Outbox Event Router to SQL Outbox Event Router's
+        Map<String, ?> convertedConfigMap = convertConfigMap(configMap);
+
+        eventRouterDelegate.configure(convertedConfigMap);
+    }
+
+    /**
+     * Replaces <i>after</i> field by parsing and expanding original JSON string to Struct type.
+     *
+     * @param originalRecord an original Record from MongoDB Connector
+     * @return a new Record of which <i>after</i> field is replaced with new one
+     * @throws Exception if <i>after</i> field of original Record is not an expected form
+     */
+    private R expandAfterField(R originalRecord) throws IllegalStateException {
```

Comment:
So what's the reasoning behind expanding the after field? The relational outbox event routing SMT puts this behind [an option](https://debezium.io/documentation/reference/stable/transformations/outbox-event-router.html#outbox-event-router-property-debezium-expand-json-payload), so I think the same should happen here.

I think I am still missing the larger picture here. Could you perhaps describe in a few sentences the entire transformation logic applied by this SMT?

Reply:
As you know, the relational outbox event routing SMT's `debezium.expand.json.payload` option is used for expanding payload field, not for after field, which includes payload field. **expandAfterField()** in MongoDB's SMT expands after field, and payload field is not expanded if configuration is not set.
The reason why I put the logic for expanding payload field in MongoDB Outbox Event Router is, I thought that the expansion of payload field in MongoDB Outbox Event Router can't be done properly in the existing relational one because JSON String format converted from BSON is different from the expected one in relational SMT. So, I thought the expansion mechanism should be different as well.

Reply:
Ok, I see (indeed I was confused between after and payload). Isn't this duplicating though what we already do in this [existing SMT](https://debezium.io/documentation/reference/stable/transformations/mongodb-event-flattening.html)? Couldn't a user achieve the same by applying the two SMTs in sequence?

Reply:
As you said, those two SMTs obviously seem to share some of the same job(expanding JSON string in after field which was originally a BSON), so I think we might be able to change the code to use *MongoDB New Document State Extraction* first, and then *MongoDB Outbox Event Router*(As you can see, we can't use these in reverse order because the format of the output record of *MongoDB Outbox Event Router* differs from what *MongoDB New Document State Extraction* requires). However, I think we need to discuss a few points. The following points are those which just come to mind at the moment.
First, the [documentation](https://debezium.io/documentation/reference/stable/transformations/mongodb-event-flattening.html) of *MongoDB New Document State Extraction* says "The SMT should be applied on a sink connector.". If it is true, I don't think we can use these two SMTs in sequence because *MongoDB Outbox Event Router* is for source connectors. If it isn't, let's just fix the documentation.
Second, *MongoDB Outbox Event Router* gets to have a strong dependency on *MongoDB New Document State Extraction* because we always have to use *MongoDB New Document State Extraction* before *MongoDB Outbox Event Router*. Would it be a good design?
Additionally, if we always have to use *MongoDB New Document State Extraction* before *MongoDB Outbox Event Router*, we always have no choice but to expand payload field, which is different from the feature of relational outbox event router. But it might not be a big deal. Of course, we might be able to add a new configuration option to *MongoDB New Document State Extraction* to exclude payload field for expansion, but it doesn't feel like a good idea in terms of the purpose of the SMT, IMHO.

Reply:
As you said, those two SMTs obviously seem to share some of the same job(expanding JSON string in after field which was originally a BSON), so I think we might be able to change the code to use *MongoDB New Document State Extraction* first, and then *MongoDB Outbox Event Router*(As you can see, we can't use these in reverse order because the format of the output record of *MongoDB Outbox Event Router* differs from what *MongoDB New Document State Extraction* requires). However, I think we need to discuss a few points. The following points are those which just come to mind at the moment.
First, the [documentation](https://debezium.io/documentation/reference/stable/transformations/mongodb-event-flattening.html) of *MongoDB New Document State Extraction* says "The SMT should be applied on a sink connector.". If it is true, we obviously can't use these two SMTs in sequence because *MongoDB Outbox Event Router* is for a source connector. If it isn't, let's just fix the documentation.
Second, *MongoDB Outbox Event Router* gets to have a strong dependency on *MongoDB New Document State Extraction* because we always have to use *MongoDB New Document State Extraction* before *MongoDB Outbox Event Router*. Would it be a good design?
Additionally, if we always have to use *MongoDB New Document State Extraction* before *MongoDB Outbox Event Router*, we always have no choice but to expand payload field, which is different from the feature of relational outbox event router. But it might not be a big deal. Of course, we might be able to add a new configuration option to *MongoDB New Document State Extraction* to exclude payload field for expansion, but it doesn't feel like a good idea in terms of the purpose of the SMT, IMHO.

Reply:
As you said, those two SMTs obviously seem to share a bit of the same job(expanding JSON string in after field which was originally a BSON). So I think we might be able to change the code to use *MongoDB New Document State Extraction* first and then *MongoDB Outbox Event Router*(I don't think we can use these in reverse order because the format of the output record of *MongoDB Outbox Event Router* differs from what *MongoDB New Document State Extraction* requires). However, I think we need to discuss a few points. The following points are those which just come to mind at the moment.
First, the [documentation](https://debezium.io/documentation/reference/stable/transformations/mongodb-event-flattening.html) of *MongoDB New Document State Extraction* says "The SMT should be applied on a sink connector". If it is true, we obviously can't use these two SMTs in sequence because *MongoDB Outbox Event Router* is for a source connector. If it isn't, let's just fix the documentation.
Second, *MongoDB New Document State Extraction* not only transforms insert operation, but also delete and update operation even though Outbox Event Router should handle them separately, but they cannot distinguish which operation an incoming record previously had because the *MongoDB New Document State Extraction* eliminates several fields that are necessary for Outbox Event Router such as "op" and "ts_ms".
Additionally, *MongoDB Outbox Event Router* gets to have a strong dependency on *MongoDB New Document State Extraction* because we always have to use *MongoDB New Document State Extraction* before *MongoDB Outbox Event Router*. Would it be a good design?

Reply:
As you said, those two SMTs obviously seem to share a bit of the same job(in terms of expanding JSON string in "after" field which was originally a BSON). So I think we might be able to change the code to use *MongoDB New Document State Extraction* first and then *MongoDB Outbox Event Router*(I don't think we can use these in reverse order because the format of the output record of *MongoDB Outbox Event Router* differs from what *MongoDB New Document State Extraction* requires). However, I think we need to discuss a few points. I have something that just comes to mind at the moment.

First, the [documentation](https://debezium.io/documentation/reference/stable/transformations/mongodb-event-flattening.html) of *MongoDB New Document State Extraction* says "The SMT should be applied on a sink connector". If it is true, we obviously can't use these two SMTs in sequence because *MongoDB Outbox Event Router* is for a source connector. If it isn't, let's just fix the documentation.

Another thing is, *MongoDB Outbox Event Router* might have to convert extracted payload struct back to JSON string for supporting `debezium.expand.json.payload` option, which might not be a big deal.

Additionally, *MongoDB Outbox Event Router* gets to have a strong dependency on *MongoDB New Document State Extraction* because we always would have to use *MongoDB New Document State Extraction* with specific configurations before *MongoDB Outbox Event Router*. For example, we always have to set *MongoDB New Document State Extraction* not to eliminate some fields, such as "op" and "ts_ms", which are necessary for *MongoDB Outbox Event Router* with `add.fields` option. I'm not sure it is a good design. Every user should always use those two SMTs, exactly understand the right settings of *MongoDB New Document State Extraction*, and also keep track of future updates both of those two SMTs for only using Outbox Pattern feature. IMHO, I'm not sure it scales. It feels more comfortable and straightforward to me to use only *MongoDB Outbox Event Router* for Outbox Pattern feature.

Reply:
As you said, I think we can consider two options to use `MongoDB New Document State Extraction` and `MongoDB Outbox Event Router` together to achieve outbox event routing in MongoDB.

* To set Kafka Connect to use the two SMTs in sequence
In this case, we must configure `MongoDB New Document State Extraction` to leave a few fields, such as `op` and `ts_ms`, which is necessary to `MongoDB Outbox Event Router`. But, I'm not sure it is a good design because the latter SMT gets to have a strong dependency on the former SMT if the order must be kept and users are forced to set specific configurations. I think it is not comfortable and straightforward to always use these two separate SMTs with specific configuration and order only to achieve Outbox Pattern in MongoDB.

* To use `MongoDB New Document State Extraction` inside `MongoDB Outbox Event Router` class
It is a similar way to using `ActivateTracingSpan` SMT in Relational Outbox SMT. I think it is a more considerable option. However, we still must write additional codes for postprocessing. For example, `MongoDB New Document State Extraction` always expands the entire `after` field and it doesn't support any configuration option to exclude specific fields from expansion. So, in order to support `debezium.expand.json.payload` for `MongoDB Outbox Event Router` we should write codes for turning the expanded payload field back to JSON string. That is, it doesn't feel particularly better than the current implementation because, after all, we should write additional codes for postprocessing and building new Schema and Struct.

IMHO, the current `MongoDB Outbox Event Router` implementation uses the same `MongoDataConverter` that is used to expand `after` field in `MongoDB New Document State Extraction`, so I don't think there's that much code duplication. Even if the code is modified to use `MongoDB New Document State Extraction`, I think the amount of new code that needs to be added will not be much different from now.

BTW, the [documentation](https://debezium.io/documentation/reference/stable/transformations/mongodb-event-flattening.html) of *MongoDB New Document State Extraction* says "The SMT should be applied on a sink connector". Is it a mistake?

Reply:
@sgc109, ok, got it now; this all makes sense, thanks a lot for the thorough explanation. I agree it's better to make this aspect of expanding the "after" part in the way as it's needed for outbox routing an implementation detail of this new SMT, opaque to the user.

> BTW, the documentation of MongoDB New Document State Extraction says "The SMT should be applied on a sink connector". Is it a mistake?

No, I don't think it is. The idea is to have the original change event in Kafka (with all the additional metadata) and then apply that SMT to those sink connectors where it make sense. It's not "wrong" to apply it to the source connector right away, but I think as a general recommendation applying it to sinks is preferable.